### PR TITLE
Feature: Use improved pluralization

### LINKED ISSUE #252: Support older version of list query name
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-codegen/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-codegen/blob/master/CONTRIBUTING.md#bugs).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### How did you install the Amplify CLI?

_No response_

### If applicable, what version of Node.js are you using?

_No response_

### Amplify CLI Version

6.1.1

### What operating system are you using?

Mac

### Amplify Codegen Command

codegen models

### Describe the bug

Please see https://github.com/aws-amplify/amplify-cli/issues/8350

Codegen changes required
https://github.com/aws-amplify/amplify-cli/issues/8350#issuecomment-933886532

### Expected behavior

We need to have backwards compatibility with the previous way to generate a plural name for list query

`amplify codegen models` can generate schema with `listPluralName` and `syncPluralName`
```
schema
   @deprecated pluralName
   listPluralName
   syncPluralName
```

Library code can use this information if it exists, otherwise, fallback to `pluralName`
```
code
    list : graphQLName = listPluralName ?? name + "s"
    sync: graphQLName = syncPluralName ?? pluralName ?? name + "s"
```

- new developers with the feature flag enabled will model gen the schema with `listPluralName` and `syncPluralName`
- existing developers will continue to use `pluralName` for sync query and `modelName + "s”` for list query. When they upgrade, the model gen will generate only `listPluralName`  and `syncPluralName`. 

We'll need to make the change in the library first, before shipping codegen changes

iOS: https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/Core/AWSPluginsCore/Model/Support/ModelSchema%2BGraphQL.swift#L19-L22

Android: https://github.com/aws-amplify/amplify-android/blob/main/aws-api-appsync/src/main/java/com/amplifyframework/api/aws/AppSyncGraphQLRequest.java#L159-L161

### Reproduction steps

Please see https://github.com/aws-amplify/amplify-cli/issues/8350

### GraphQL schema(s)

_No response_

### Log output

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by marcvberg**
#### Description of changes

Use feature flag to determine whether to set pluralName or listPluralName/syncPluralName

#### Issue #, if available

https://github.com/aws-amplify/amplify-cli/issues/8350
fix #252 

#### Description of how you validated changes
Unit testing

Manually tested with JS, Android, and iOS applications. Manual tests involved creating Todos through the console as well as through the app, and observing that the data was still available while offline

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-codegen/blob/master/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/appsync-modelgen-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap [hunk 1]
```diff
@@ -39,29 +39,29 @@ public final class customClaim implements Model {
   public String getId() {
       return id;
   }
-
```
Comment (dpilch) [inline packages/appsync-modelgen-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap:42]: It looks like your editor made auto-formatting changes to the snapshots. Could you revert these changes?

**Review by dpilch — CHANGES_REQUESTED** [review]

File: packages/appsync-modelgen-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap (hunk 1 above)
Comment (marcvberg) [reply packages/appsync-modelgen-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap]: I think I got it fixed - waiting on tests to run

**Review by dpilch — APPROVED** [review]

**Review by AaronZyLee — COMMENTED** [review]
The codegen logic LGTM. Still I have some concerns on each library change, which need further confirmation with library side.

- JS(meta): the new attributes are appended so it should not break customers

- iOS and Android: the original plural field is replaced by two new fields, need library side release first

- Flutter: we do not receive ticket from flutter and it usually depends on iOS and Android team

Also need a decision on the default feature flag value

**Review by manueliglesias — DISMISSED** [review]
LGTM

JS team is OK with this change

File: packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts [hunk 2]
```diff
@@ -610,13 +611,18 @@ export class AppSyncModelDartVisitor<
   }
 
   protected generateSchemaField(model: CodeGenModel, declarationBlock: DartDeclarationBlock): void {
+    const pluralFields = this.config.improvePluralization ? `modelSchemaDefinition.listPluralName = "${
```
Comment (AaronZyLee) [inline packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts:614]: As mentioned in the comment, Flutter need option 2 for the change (having all three fields instead of replacing)

Comment (HuiSF) [reply packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts:614]: **Correction:** amplify-flutter should take the **option 1**, the dart-visitor should not make any changes.

**Comment by lawmicha** [comment]
Reminder this PR needs to be updated to remove the dart files changes related to Flutter (@HuiSF's comment https://github.com/aws-amplify/amplify-codegen/pull/255#discussion_r733981564)

File: packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts (hunk 2 above)
Comment (marcvberg) [reply packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts]: @AaronZyLee removed

File: packages/appsync-modelgen-plugin/src/visitors/appsync-json-metadata-visitor.ts [hunk 3]
```diff
@@ -189,11 +192,17 @@ export class AppSyncJSONVisitor<
     }));
   }
   private generateModelMetadata(model: CodeGenModel): JSONSchemaModel {
+    const pluralAttributes = this.config.improvePluralization ?
+      {
+        listPluralName: plurality(model.name, this.config.improvePluralization),
+        syncPluralName: this.pluralizeModelName(model)
```
Comment (lazpavel) [inline packages/appsync-modelgen-plugin/src/visitors/appsync-json-metadata-visitor.ts:198]: should json syncPluralName be different than java visitor? 
https://github.com/aws-amplify/amplify-codegen/pull/255/files#diff-8fdb1b5aed85eca745295520073edb2af47d27fdf8cff3173487f147b11e5fdaR761

Comment (marcvberg) [reply packages/appsync-modelgen-plugin/src/visitors/appsync-json-metadata-visitor.ts:198]: good catch I think the java visitor needs fixed

**Review by lazpavel — DISMISSED** [review]
LGTM

**Comment by sebsto** [comment]
hello, any chance to get this in the next CLI release ?

**Review by AaronZyLee — APPROVED** [review]
lgtm 🐰

**Comment by lawmicha** [comment]
any updates on this?

**Comment by lawmicha** [comment]
Another variation of this problem exists in V2 transform, if V2 uses the improved pluralization by default for both list and sync query -> https://github.com/aws-amplify/amplify-ios/issues/1707#issuecomment-1079413894

**Comment by alharris-at** [comment]
Outstanding work on this PR. Confirm we are clear to push this for all platforms, and then rebase against current `master` branch.

**Comment by alharris-at** [comment]
Hi @marcvberg could you reopen this PR against the `main` branch?

**Comment by alharris-at** [comment]
@marcvberg let's create an asana task to get this released in the next sprint.

**Comment by marcvberg** [comment]
@alharris-at Added to my tasks

**Comment by sebsto** [comment]
Hello, Any status when this will be merged. It blocks customer projects on this issue https://github.com/aws-amplify/amplify-ios/issues/1443 (opened since Sept. 2021). This PR is open since Oct 2021. Can you help to prioritize this ?

**Review by AaronZyLee — COMMENTED** [review]
The failing tests are probably due to the dependency change. You can rebase the main branch (or merge main) with the latest dependency versions and re-build the project.

File: packages/appsync-modelgen-plugin/src/__tests__/visitors/appsync-swift-visitor.test.ts [hunk 4]
```diff
@@ -2279,6 +2280,72 @@ describe('AppSyncSwiftVisitor', () => {
     });
   });
 
+  describe('Use Improved Pluralization', () => {
+    let wishSchema: string;
+    beforeEach(() => {
+      wishSchema = /* GraphQL */ `
```
Comment (AaronZyLee) [inline packages/appsync-modelgen-plugin/src/__tests__/visitors/appsync-swift-visitor.test.ts:2286]: Nit: We can use the schema here to display the difference before and after the plural name change. Refer https://github.com/aws-amplify/amplify-cli/issues/8350
```
type NoteData
@model
{
    id: ID!
    name: String!
    description: String
    image: String
}

type Wishes @model 
@key(name: "byUsers", fields: ["idUser"]) 
{

  id: ID!
  date: AWSDateTime!
  idUser: ID!
  ev: ID! 
  deprecated: Boolean!  
  owner: String         
}

type Wish @model 
@key(name: "byUsers", fields: ["idUser"]) {

  id: ID!
  date: AWSDateTime!
  idUser: ID!
  ev: ID! 
  deprecated: Boolean!  
  owner: String         
}
```

**Review by AaronZyLee — CHANGES_REQUESTED** [review]
Could you add a unit test for the `schema.js` generated by JS modelgen since the output is changed? Otherwise LGTM

File: package.json [hunk 5]
```diff
@@ -4,7 +4,7 @@
   "description": "amplify-codegen",
   "scripts": {
     "test-changed": "lerna run test --since main",
-    "test": "lerna run test",
+    "test": "lerna run test -- -u",
```
Comment (AaronZyLee) [inline package.json:7]: This change will override the snapshots in CI without commits and should be removed

**Review by AaronZyLee — CHANGES_REQUESTED** [review]

File: package.json (hunk 5 above)
Comment (marcvberg) [reply package.json]: Fixed

**Review by AaronZyLee — DISMISSED** [review]
LGTM. This will be held until the studio CMS using the new model introspection schema as there are new fields introduced in `schema.js`

**Comment by alharris-at** [comment]
@sebsto what platform are you writing your workshop for? We're testing this right now to verify if we're in a good place to merge, but it looks like Android/Flutter still won't have support for improvedPluralization after this PR from my current testing.

**Comment by sebsto** [comment]
@alharris-at workshop is for iOS / Swift 
I currently have instructions guiding customers to modify the generated code 

(and while I am at it, there is a warning about field ID being deprecated. but I should report that on another issue)

**Comment by lawmicha** [comment]
> @alharris-at workshop is for iOS / Swift I currently have instructions guiding customers to modify the generated code
> 
> (and while I am at it, there is a warning about field ID being deprecated. but I should report that on another issue)

the warning about field ID being deprecated is a swift codegen warning. This is related to the new `identifier` field that is being created for the ability to create customizations of the primary key (renaming / composite, etc). To enable this, you can provide some instructions to the developer to toggle `respectprimarykeyattributesonconnectionfield` to true in the `cli.json` file before provisioning the backend (amplify push) and codegenerating the swift models (`amplify codegen models`)

**Comment by afern247** [comment]
The last comment on this PR was on October. Is someone actively working on this? I have an open issue reported: https://github.com/aws-amplify/amplify-swift/issues/2460

**Comment by sebsto** [comment]
Thank you @marcvberg for the push. What are the next steps to get this merged ?

**Review by AaronZyLee — APPROVED** [review]

File: packages/appsync-modelgen-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap [hunk 6]
```diff
@@ -23,7 +23,7 @@ import static com.amplifyframework.core.model.query.predicate.QueryField.field;
 
 /** This is an auto generated class representing the Post type in your schema. */
 @SuppressWarnings(\\"all\\")
-@ModelConfig(pluralName = \\"Posts\\", type = Model.Type.USER, version = 1)
+@ModelConfig(listPluralName = \\"Posts\\", syncPluralName = \\"Posts\\", type = Model.Type.USER, version = 1)
```
Comment (alharris-at) [inline packages/appsync-modelgen-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap:26]: Is this a breaking change to no longer also specify the 'pluralName'?

File: packages/appsync-modelgen-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap [hunk 7]
```diff
@@ -1845,7 +1845,7 @@ import static com.amplifyframework.core.model.query.predicate.QueryField.field;
 
 /** This is an auto generated class representing the Post type in your schema. */
 @SuppressWarnings(\\"all\\")
-@ModelConfig(pluralName = \\"Posts\\")
```
Comment (alharris-at) [inline packages/appsync-modelgen-plugin/src/__tests__/visitors/__snapshots__/appsync-java-visitor.test.ts.snap:1848]: Why do some of these not have the type/version as well?

**Review by alharris-at — COMMENTED** [review]
Q on the interface, otherwise looks sane to me given we understand the e2e testing a bit more.

**Comment by sebsto** [comment]
Looks like this issue has been fixed by https://github.com/aws-amplify/amplify-swift/issues/3135
